src/fortran/configoptions.F90

```diff
@@ -0,0 +1,254 @@
+!
+! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+! AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+! DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+! FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+module smartredis_configoptions
+
+use iso_c_binding,   only : c_ptr, c_bool, c_null_ptr, c_char, c_int
+use iso_c_binding,   only : c_int8_t, c_int16_t, c_int32_t, c_int64_t, c_float, c_double, c_size_t
+use iso_c_binding,   only : c_loc, c_f_pointer
+
+use, intrinsic :: iso_fortran_env, only: stderr => error_unit
+
+use fortran_c_interop, only : convert_char_array_to_c, enum_kind, C_MAX_STRING
+
+implicit none; private
+
+include 'enum_fortran.inc'
+include 'configoptions/configoptions_interfaces.inc'
+
+public :: enum_kind !< The kind of integer equivalent to a C enum. According to C an Fortran
+                    !! standards this should be c_int, but is renamed here to ensure that
+                    !! users do not have to import the iso_c_binding module into their
+                    !! programs
+
+!> Contains multiple tensors and metadata used to describe an entire set of data
+type, public :: configoptions_type
+  type(c_ptr) :: configoptions_ptr !< A pointer to the initialized dataset object
+
+  contains
+
+  !> Access the raw C pointer for the configoptions
+  procedure :: get_c_pointer
+
+  ! Factory methods
+  !> Instantiate ConfigOptions, getting selections from environment variables
+  procedure :: create_configoptions_from_environment
+  ! Instantiate ConfigOptions, getting selections from a file with JSON data
+  !procedure :: create_configoptions_from_file
+  ! Instantiate ConfigOptions, getting selections from a string containing a JSON blob
+  !procedure :: create_configoptions_from_string
+
+  ! Option access
+  !> Retrieve the value of a numeric configuration option
+  procedure :: get_integer_option
+  !> Retrieve the value of a string configuration option
+  procedure :: get_string_option
+  !> Check whether a configuration option is set
+  procedure :: is_configured
+
+  ! Option value overrides
+  !> Override the value of a numeric configuration option
+  procedure :: override_integer_option
+  !> Override the value of a string configuration option
+  procedure :: override_string_option
+
+end type configoptions_type
+
+contains
+
+
+!> Access the raw C pointer for the dataset
+function get_c_pointer(self)
```

Comment:
I don't think this function is needed here.

Reply:
It will be needed when we get to the next stage and need to get to the C pointer to pass to a Client constructor